gui/src/3D/controls/slider3D.ts

```diff
@@ -119,64 +135,108 @@ export class Slider3D extends Control3D {
-        sliderThumb.position.x = this._convertToPosition(this.value);
-        sliderThumb.addBehavior(this._createBehavior());
-        sliderThumb.setParent(anchor);
-
-        this._sliderBar = sliderBar;
+        const sliderBackplate = BoxBuilder.CreateBox(`${this.name}_sliderbackplate`, {
+            width: 1.0,
+            height: 1.0,
+            depth: 1.0,
+        }, scene);
+        sliderBackplate.isPickable = false;
+        sliderBackplate.visibility = 0;
+        sliderBackplate.scaling = new Vector3(1, 0.5, 0.8);
+
+        SceneLoader.ImportMeshAsync(
+            undefined,
+            Slider3D.MODEL_BASE_URL,
+            Slider3D.MODEL_FILENAME,
+            scene)
+            .then((result) => {
+                const sliderBackplateModel = result.meshes[1];
+                const sliderBarModel = result.meshes[1].clone(`${this.name}_sliderbar`, sliderBackplate);
+                const sliderThumbModel = result.meshes[1].clone(`${this.name}_sliderthumb`, sliderBackplate);
+                sliderBackplateModel.visibility = 0;
+
+                if (this._sliderBackplateVisible) {
+                    sliderBackplateModel.visibility = 1;
+                    sliderBackplateModel.name = `${this.name}_sliderbackplate`;
+                    sliderBackplateModel.isPickable = false;
+                    sliderBackplateModel.scaling.x = 1;
+                    sliderBackplateModel.scaling.z = 0.2;
+                    sliderBackplateModel.parent = sliderBackplate;
+                    if (!!this._sliderBackplateMaterial) {
+                        sliderBackplateModel.material = this._sliderBackplateMaterial;
+                    }
+                    this._sliderBackplate = sliderBackplateModel;
+                }
+
+                if (!!sliderBarModel) {
+                    sliderBarModel.parent = sliderBackplate;
+                    sliderBarModel.position.z = -0.1;
+                    sliderBarModel.scaling = new Vector3(SLIDER_SCALING - SLIDER_MARGIN, 0.04, 0.3);
+                    sliderBarModel.isPickable = false;
+                    if (!!this._sliderBarMaterial) {
+                        sliderBarModel.material = this._sliderBarMaterial;
+                    }
+                    this._sliderBar = sliderBarModel;
+                }
+
+                if (!!sliderThumbModel) {
+                    sliderThumbModel.isPickable = true;
+                    sliderThumbModel.position.z = -0.115;
+                    sliderThumbModel.scaling = new Vector3(0.025, 0.3, 0.6);
+                    sliderThumbModel.position.x = this._convertToPosition(this.value);
+                    sliderThumbModel.addBehavior(this._createBehavior());
+                    if (!!this._sliderThumbMaterial) {
+                        sliderThumbModel.material = this._sliderThumbMaterial;
+                    }
+                    this._sliderThumb = sliderThumbModel;

```

Comment:
This is being set in line 231 already. It is then re-set here. Is there a need to set it in lines 229-231?
And if it needed, you will need to dispose the box that was created in 229.

Reply:
```suggestion
                    if(this._sliderThumb) {
                        this._sliderThumb.dispose();
                    }
                    this._sliderThumb = sliderThumbModel;
```

Reply:
@RaananW I like your suggestion. However, it doesn't solve the original reason on why I need this workaround. 

What I have noticed is that without this workaround, the models get rendered without materials. I suspect that this is associated with [`get mesh()` ](https://github.com/BabylonJS/Babylon.js/blob/81be3c37822d53c9cd38d50c29a1243791eae607/gui/src/3D/controls/slider3D.ts#L70) because when that is disabled, the materials exist. However, this creates a new bug where the `SliderThumb ` doesn't have the associated behavior.

Do you have any idea about this? 

![image](https://user-images.githubusercontent.com/15232740/134346305-83b552af-2c8e-44e8-9875-f42c0e839d20.png)

Reply:
Would be great if you could show me the issue, maybe we can resolve it together. It's hard for me to understand the lifecycle of those meshes.
If I understand correctly this mesh should only render when the material is ready? you can hide it until then (isVisible = true), or only set it after the material is ready (i.e. don't set the first this._sliderMesh at all)

Reply:
I was looking at the Control3D class https://github.com/BabylonJS/Babylon.js/blob/a32c04e6a9c50e17ca9bc9a1834aa7a546c23bb8/gui/src/3D/controls/control3D.ts#L257-L271 
What I understand is that when Line 266 is executed, the meshes from the [ slider3D's `_createNode`](https://github.com/BabylonJS/Babylon.js/blob/2dcd4da1278799b0518110575f206703a0b94837/gui/src/3D/controls/slider3D.ts#L181) are not yet loaded, `affectMaterial` on line 270 is never reached and the material doesn't get applied when the meshes eventually load.

Another approach I tried was to use the [Control3D's `get mesh()`](https://github.com/BabylonJS/Babylon.js/blob/a32c04e6a9c50e17ca9bc9a1834aa7a546c23bb8/gui/src/3D/controls/control3D.ts#L236-L242), while this approach reaches `affectMaterial` on line 270, it returns a `TransformNode` which is pickable on line 268 but makes the sliderMesh unpickable.

I will be glad to walk you through this issues. We can sync on this at your convenience.

Reply:
I see the issue now.
`_createNode` is expected to be synchronous, but is async when loading external meshes, which is not supported by the current construction architecture.
But it seems to be solved in the holographic button almost the same way as you do. Why is it important to initialize `this._sliderThumb` during construction? isn't it enough if it is available after the model has finished loading?
`_affectMaterial` is creating the materials. You can check if they are available in this function and then apply it to them, and if not, apply them after the model has loaded. 
this way _createNode returns a mesh (the collider box), so affectMaterial is being executed correctly. then, after the models were loaded, you can set the materials. Holographic backplate, though simpler than this class, can show you the way to solve this.